subsys/bluetooth/audio/bap_broadcast_assistant.c

```diff
@@ -184,14 +191,159 @@ static int parse_recv_state(const void *data, uint16_t length,
+					 uint16_t length,
+					 struct bt_bap_scan_delegator_recv_state *recv_state)
+{
+	int err;
+	int16_t index;
+
+	err = parse_recv_state(data, length, recv_state);
+
+	if (err != 0) {
+		LOG_WRN("Invalid receive state received");
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	index = lookup_index_by_handle(broadcast_assistant.long_read_handle);
+	if (index < 0) {
+		LOG_DBG("Invalid index");
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	broadcast_assistant.src_ids[index] = recv_state->src_id;
+	broadcast_assistant.past_avail[index] = past_available(conn,
+							       &recv_state->addr,
+							       recv_state->adv_sid);
+
+	if (broadcast_assistant_cbs != NULL &&
+	    broadcast_assistant_cbs->recv_state != NULL) {
+		broadcast_assistant_cbs->recv_state(conn, 0, recv_state);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static uint8_t broadcast_assistant_bap_ntf_read_func(struct bt_conn *conn, uint8_t err,
+						     struct bt_gatt_read_params *read,
+						     const void *data, uint16_t length)
+{
+	struct bt_bap_scan_delegator_recv_state recv_state;
+	uint16_t handle = read->single.handle;
+	uint16_t data_length;
+
+	LOG_DBG("conn %p err 0x%02x len %u", conn, err, length);
+
+	if (err) {
+		LOG_DBG("Failed to read: %u", err);
+		memset(read, 0, sizeof(*read));
+		bap_ntf_read_done();
+		bt_conn_unref(conn);
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	LOG_DBG("handle 0x%04x", handle);
+
+	if (data != NULL) {
+		if (net_buf_simple_tailroom(&cp_buf) < length) {
+			LOG_DBG("Buffer full, invalid server response of size %u",
+				length + cp_buf.len);
+			memset(read, 0, sizeof(*read));
+			bap_ntf_read_done();
+			bt_conn_unref(conn);
+
+			return BT_GATT_ITER_STOP;
+		}
+
+		/* store data*/
+		net_buf_simple_add_mem(&cp_buf, data, length);
+
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	/*
+	 * we reset the buffer so that it is ready for new data
+	 */
+	memset(read, 0, sizeof(*read));
+	data_length = cp_buf.len;
+	bap_ntf_read_done();
+
+	/* do the parse and callback to send  notify to application*/
+	parse_and_send_recv_state(conn, cp_buf.data, data_length, &recv_state);
+
+	bt_conn_unref(conn);
+
+	return BT_GATT_ITER_STOP;
+}
+
+
+static void long_bap_read(struct bt_conn *conn, uint16_t handle)
+{
+	int err;
+
+	LOG_DBG("conn %p busy %u", conn, broadcast_assistant.busy);
+
+	if (broadcast_assistant.busy) {
+		/* If the client is busy reading or writing something else, reschedule the
+		 * long read.
+		 */
+		struct bt_conn_info conn_info;
+
+		err = bt_conn_get_info(conn, &conn_info);
+		if (err != 0) {
+			LOG_DBG("Failed to get conn info, use default interval");
+
+			conn_info.le.interval = BT_GAP_INIT_CONN_INT_MIN;
+		}
+
+		/* Wait a connection interval to retry */
+		err = k_work_reschedule(&broadcast_assistant.bap_read_work,
+					K_USEC(BT_CONN_INTERVAL_TO_US(conn_info.le.interval)));
+		if (err < 0) {
+			LOG_DBG("Failed to reschedule read work: %d", err);
+			bap_ntf_read_done();
+			bt_conn_unref(conn);
```

Comment:
Ditto with the `unref`

I think it makes sense to do 1 `bt_conn_ref` when you initate the discover and store the `conn` pointer in the instance, and then 1 `bt_conn_unref` when you disconnect or the discover function is called again. 

See e.g. the similar changes here: https://github.com/zephyrproject-rtos/zephyr/pull/67629/files#diff-a1827bc1ecbe96b568cf0cc5ea28bf55f8593f2ae005a9c65f6f051254b21954